csi/node.go

```diff
@@ -140,13 +140,71 @@ func (s *OsdCsiServer) NodePublishVolume(
 			err.Error())
 	}
 
-	dlog.Infof("Voume %s mounted on %s",
+	dlog.Infof("Volume %s mounted on %s",
 		req.GetVolumeId(),
 		req.GetTargetPath())
 
 	return &csi.NodePublishVolumeResponse{}, nil
 }
 
+// NodeUnpublishVolume is a CSI API call which unmounts the volume.
+//
+// TODO: Support READ ONLY Mounts
+//
+func (s *OsdCsiServer) NodeUnpublishVolume(
+	ctx context.Context,
+	req *csi.NodeUnpublishVolumeRequest,
+) (*csi.NodeUnpublishVolumeResponse, error) {
+
+	dlog.Debugf("NodeUnPublishVolume req[%#v]", req)
+
+	// Check arguments
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	}
+	if len(req.GetTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path must be provided")
+	}
+
+	// Get volume information
+	v, err := util.VolumeFromName(s.driver, req.GetVolumeId())
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "Volume id %s not found: %s",
+			req.GetVolumeId(),
+			err.Error())
+	}
+
+	// Verify target location is an existing directory
+	// See: https://github.com/container-storage-interface/spec/issues/60
+	if err = verifyTargetLocation(req.GetTargetPath()); err != nil {
+		return nil, status.Error(codes.NotFound, err.Error())
+	}
+
+	// Mount volume onto the path
+	if err = s.driver.Unmount(v.GetId(), req.GetTargetPath(), nil); err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to unmount volume %s onto %s: %s",
+			req.GetVolumeId(),
+			req.GetTargetPath(),
+			err.Error())
+	}
+
+	if err = s.driver.Detach(req.GetVolumeId(), nil); err != nil {
```

Comment:
We should only do deatch for Type() Block, afaik

Reply:
@adityadani is that correct? if so i'll make the fix

Reply:
Adding a check here similar to one in NodePublishVolume () means csi code would need to change if a driver implementation changes. This could be prevented by appropriate handling of unimplemented calls in the driver.

Reply:
I agree with @prabirpaul . All the drivers in openstorage do implement Detach calls. It could just be a no-op for such drivers. Ideally we won't need this extra check.

Reply:
That's fine, but then we would need to create a new issue for this to solve at another PR.

Reply:
So keep the check for block or not? my last commit has this check.